### DISCUSSION TITLE
Approach to the maintenance of Local SNOMED CT Extensions

### DISCUSSION THREAD

**Discussion by martingall87** [b2ihealthcare/snow-owl#1038] [Q&A] [answered]
Hi,

I'm just experimenting with Snow Owl (8.3.0), looking at how it could be used to maintain a local extensions in SNOMED CT, specifically local extensions of the SNOMED CT UK Edition.

In the SNOMED CT UK Edition there are 4 main components to the release files that ultimately depend on the International Edition. I’ve illustrated the desired scenario below:

```
International
  | - UK Edition
        | - UK Clinical
        |     | - UK Clinical Reference Sets
	|     | - ** Local Clinical Extension **
        |
        | - UK Drug
	       | - ** Local Drug Extension **
```

I've taken the following steps to import a base version of the UK Edition / UK Drug Extension (32.12):

1. Import International
- Create codesystem 'SNOMEDCT' via `POST /codesystems`
- Import files via `POST /snomedct/SNOMEDCT/import`
- Create version 'SNOMEDCT/2021-01-31' via `POST /versions`

2. Import UK Edition 
- Create codesystem 'SNOMEDCT-UK' as extension of 'SNOMEDCT/2021-01-31' via `POST /codesystems`
- Import files via `POST /snomedct/SNOMEDCT-UK/import`
- Create version SNOMEDCT-UK/2022-04-13 via `POST /versions`

3. Import UK Drug Extension
- Create codesystem 'SNOMEDCT-UK-DR' as extension of 'SNOMEDCT-UK/2022-04-13'  via `POST /codesystems`
- Import files via `POST /snomedct/SNOMEDCT-UK-DR/import`
- Create version SNOMEDCT-UK-DR/2022-04-13 via `POST /versions`

Finally create a local extension as an extension of UK Drug Extension 'SNOMEDCT-DR-UK/2022-04-13'

I’d now like to follow the process for upgrading to a later version of the UK Edition / UK Drug Extension (33.1) which also requires an uplift of the International version too:

- International 2021-07-31
- UK Edition 33.1 (2022-06-08)
- UK Drug 33.1 (2022-06-08)

To import a later version of the International I believe this can be done via the following steps:
- Import files via `POST /snomedct/SNOMEDCT/import`
- Create version SNOMEDCT/2021-07-31 via `POST /versions`

What would be the correct workflow to upgrade the dependant code systems (UK Edition, UK Drug Extension & Local Extension) to align with the latest international + import the latest content from the 33.1.0 (2022-06-08) release?

Would the approach be to...
1. Use POST /upgrade to start the upgrade for UK Edition
2. Import the latest 33.1 content into the ‘UPGRADE’ codesystem
3. Use POST /upgrade/{id}/complete to complete the upgrade
4. Create a new version for UK Edition (33.1.0 2022-06-08)

Then repeat the above steps for UK Drug Extension and Local extension?

Many thanks!

**Comment by cmark** [comment, marked as answer]
Hi @martingall87,

> I've taken the following steps to import a base version of the UK Edition / UK Drug Extension (32.12):

These look good to me so far. One small note, you can use the `createVersions` parameter of the RF2 import to automatically generate the versions for you, but of course doing it manually is fine as well.

> To import a later version of the International I believe this can be done via the following steps:
> * Import files via POST /snomedct/SNOMEDCT/import
> * Create version SNOMEDCT/2021-07-31 via POST /versions

Yes, using the import endpoint you can import the official Full release of the International Edition and Snow Owl will generate a Delta internally and import the supported content. Versions can be automatically generated via the `createVersions` RF2 import request body parameter or manually as you have described.

> What would be the correct workflow to upgrade the dependant code systems (UK Edition, UK Drug Extension & Local Extension) to align with the latest international + import the latest content from the 33.1.0 (2022-06-08) release?
> 
> Would the approach be to...
> 1. Use POST /upgrade to start the upgrade for UK Edition
> 2. Import the latest 33.1 content into the ‘UPGRADE’ codesystem
> 3. Use POST /upgrade/{id}/complete to complete the upgrade
> 4. Create a new version for UK Edition (33.1.0 2022-06-08)

Yes, using the deep branching strategy you have described the recommended way to upgrade Extensions to newer dependency versions is to use the Code System Upgrade API. The steps you mention above look good to me, make sure you use `createVersions=false` when importing content to Upgrade Code Systems because that feature is unsupported on such resources. After successfully completing the upgrade you can version the content using the new version tag and effective time.

The same applies to the other UK Extensions, yes.

For Local Extensions, the process is similar but if you are maintaining new SNOMED CT content in your Local Extension then you might need to perform migration of that content to the new changes introduced by the upstream dependency. Therefore before completing the Upgrade Code System of your Local Extension you have to manually update your Local concepts/descriptions/etc. to get the latest changes. You can use the Validation API to perform checks on your content automatically and aid the migration process. There are several open-source validation rules available in the OSS version [here](https://github.com/b2ihealthcare/snow-owl/blob/8.x/snomed/com.b2international.snowowl.validation.snomed/src/main/resources/validation-rules-common.json), but if you think you need more feel free to develop your own using simple queries or Groovy scripts. Or if you have a more complex use case feel free to contact us for discussing it further.

I hope this clarifies some things and helps. Good luck with the upgrades.

Cheers,
Mark
